### PR TITLE
REF: implement LossySetitemError

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

Catch this specifically instead of ValueError, which can have unrelated causes xref https://github.com/numpy/numpy/issues/20923

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -2087,3 +2089,11 @@ def _dtype_can_hold_range(rng: range, dtype: np.dtype) -> bool:
     if not len(rng):
         return True
     return np.can_cast(rng[0], dtype) and np.can_cast(rng[-1], dtype)
+
+
+class LossySetitemError(Exception):
```
Comment (mroeschke) [inline pandas/core/dtypes/cast.py:2094]: So this is not a user-exposed error?

**Comment by jbrockmendel** [comment]
That’s the idea, yes

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/dtypes/cast.py:2094]: Cool. Yeah if it was probably should be exposed in pandas.errors

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]
lgtm ping on green

**Comment by jbrockmendel** [comment]
ping(ish)

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
hmm rebase and i think some related failures

**Comment by jbrockmendel** [comment]
> hmm rebase and i think some related failures

rebased + green

**Comment by mroeschke** [comment]
@jbrockmendel looks like the doctests and ASV are failing as this error is being surfaced

https://github.com/pandas-dev/pandas/runs/5014539859?check_suite_focus=true
https://github.com/pandas-dev/pandas/runs/5014540218?check_suite_focus=true

(odd that the build didn't catch this)

**Pull request merged**